File: app/src/main/java/org/thoughtcrime/securesms/linkpreview/LinkPreviewState.java
/*
 * Copyright 2023 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.thoughtcrime.securesms.linkpreview;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Optional;

public class LinkPreviewState {
  private final String                      activeUrlForError;
  private final boolean                     isLoading;
  private final boolean                     hasLinks;
  private final Optional<LinkPreview>       linkPreview;
  private final LinkPreviewRepository.Error error;

  private LinkPreviewState(@Nullable String activeUrlForError,
                           boolean isLoading,
                           boolean hasLinks,
                           Optional<LinkPreview> linkPreview,
                           @Nullable LinkPreviewRepository.Error error)
  {
    this.activeUrlForError = activeUrlForError;
    this.isLoading         = isLoading;
    this.hasLinks          = hasLinks;
    this.linkPreview       = linkPreview;
    this.error             = error;
  }

  public static LinkPreviewState forLoading() {
    return new LinkPreviewState(null, true, false, Optional.empty(), null);
  }

  public static LinkPreviewState forPreview(@NonNull LinkPreview linkPreview) {
    return new LinkPreviewState(null, false, true, Optional.of(linkPreview), null);
  }

  public static LinkPreviewState forLinksWithNoPreview(@Nullable String activeUrlForError, @NonNull LinkPreviewRepository.Error error) {
    return new LinkPreviewState(activeUrlForError, false, true, Optional.empty(), error);
  }

  public static LinkPreviewState forNoLinks() {
    return new LinkPreviewState(null, false, false, Optional.empty(), null);
  }

  public @Nullable String getActiveUrlForError() {
    return activeUrlForError;
  }

  public boolean isLoading() {
    return isLoading;
  }

  public boolean hasLinks() {
    return hasLinks;
  }

  public Optional<LinkPreview> getLinkPreview() {
    return linkPreview;
  }

  public @Nullable LinkPreviewRepository.Error getError() {
    return error;
  }

  public boolean hasContent() {
    return isLoading || hasLinks;
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/linkpreview/LinkPreviewState.kt
/*
 * Copyright 2023 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */
package org.thoughtcrime.securesms.linkpreview

import android.os.Parcelable
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import java.util.Optional

@Parcelize
class LinkPreviewState private constructor(
  @JvmField val activeUrlForError: String?,
  @JvmField val isLoading: Boolean,
  private val hasLinks: Boolean,
  private val preview: LinkPreview?,
  @JvmField val error: LinkPreviewRepository.Error?,
  @JvmField val link: Link?
) : Parcelable {

  @IgnoredOnParcel
  @JvmField
  val linkPreview: Optional<LinkPreview> = Optional.ofNullable(preview)

  fun hasLinks(): Boolean {
    return hasLinks
  }

  fun hasContent(): Boolean {
    return isLoading || hasLinks
  }

  val url: String? = link?.url ?: preview?.url ?: activeUrlForError

  companion object {
    @JvmStatic
    fun forLoading(link: Link): LinkPreviewState {
      return LinkPreviewState(
        activeUrlForError = null,
        isLoading = true,
        hasLinks = false,
        preview = null,
        error = null,
        link = link
      )
    }

    @JvmStatic
    fun forPreview(linkPreview: LinkPreview): LinkPreviewState {
      return LinkPreviewState(
        activeUrlForError = null,
        isLoading = false,
        hasLinks = true,
        preview = linkPreview,
        error = null,
        link = null
      )
    }

    @JvmStatic
    fun forLinksWithNoPreview(activeUrlForError: String?, error: LinkPreviewRepository.Error): LinkPreviewState {
      return LinkPreviewState(
        activeUrlForError = activeUrlForError,
        isLoading = false,
        hasLinks = true,
        preview = null,
        error = error,
        link = null
      )
    }

    @JvmStatic
    fun forNoLinks(): LinkPreviewState {
      return LinkPreviewState(
        activeUrlForError = null,
        isLoading = false,
        hasLinks = false,
        preview = null,
        error = null,
        link = null
      )
    }
  }
}
